File: backend/app/schemas/production_execution.py
from __future__ import annotations
from datetime import datetime
from decimal import Decimal
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict


# ── Work Order ─────────────────────────────────────────────────────────────────

class ExecWorkOrderCreate(BaseModel):
    step_name: str
    step_sequence: int
    planned_qty: Decimal
    setup_time_planned: Decimal = Decimal("0")
    run_time_planned: Decimal = Decimal("0")
    cleanup_time_planned: Decimal = Decimal("0")
    work_center_id: Optional[UUID] = None
    operation_code: Optional[str] = None
    section_id: Optional[str] = None
    line_id: Optional[str] = None
    qc_required_flag: bool = False
    dependency_work_order_id: Optional[UUID] = None
    notes: Optional[str] = None


class ExecWorkOrderOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    production_order_id: UUID
    operation_code: Optional[str]
    step_sequence: int
    step_name: str
    work_center_id: Optional[UUID]
    section_id: Optional[str]
    line_id: Optional[str]
    machine_id: Optional[UUID]
    assigned_operator_id: Optional[UUID]
    planned_qty: Decimal
    completed_qty: Decimal
    scrap_qty: Decimal
    rework_qty: Decimal
    setup_time_planned: Decimal
    run_time_planned: Decimal
    cleanup_time_planned: Decimal
    setup_time_actual: Optional[Decimal]
    run_time_actual: Optional[Decimal]
    cleanup_time_actual: Optional[Decimal]
    dependency_work_order_id: Optional[UUID]
    qc_required_flag: bool
    qc_passed: Optional[bool]
    scrap_category: Optional[str]
    scrap_reason: Optional[str]
    status: str
    started_at: Optional[datetime]
    completed_at: Optional[datetime]
    notes: Optional[str]
    created_at: datetime
    updated_at: datetime


class CompleteWorkOrderRequest(BaseModel):
    completed_qty: Decimal
    scrap_qty: Decimal = Decimal("0")
    rework_qty: Decimal = Decimal("0")
    setup_time_actual: Optional[Decimal] = None
    run_time_actual: Optional[Decimal] = None
    cleanup_time_actual: Optional[Decimal] = None
    scrap_category: Optional[str] = None
    scrap_reason: Optional[str] = None
    qc_passed: Optional[bool] = None
    notes: Optional[str] = None


# ── Material ───────────────────────────────────────────────────────────────────

class ExecOrderMaterialCreate(BaseModel):
    item_type: str = "MATERIAL"
    item_id: Optional[UUID] = None
    item_name: Optional[str] = None
    item_code: Optional[str] = None
    required_qty: Decimal
    uom: str
    work_order_id: Optional[UUID] = None
    unit_cost: Optional[Decimal] = None


class IssueReturnMaterialRequest(BaseModel):
    material_id: UUID
    qty: Decimal
    lot_no: Optional[str] = None
    action: str  # "issue" or "return"


class ExecOrderMaterialOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    production_order_id: UUID
    work_order_id: Optional[UUID]
    item_type: str
    item_id: Optional[UUID]
    item_name: Optional[str]
    item_code: Optional[str]
    required_qty: Decimal
    reserved_qty: Decimal
    issued_qty: Decimal
    consumed_qty: Decimal
    returned_qty: Decimal
    scrapped_qty: Decimal
    uom: str
    status: str
    lot_numbers: Optional[list]
    unit_cost: Optional[Decimal]
    extended_cost: Optional[Decimal]


# ── Genealogy ──────────────────────────────────────────────────────────────────

class BatchGenealogyOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    production_order_id: UUID
    input_lot_no: Optional[str]
    input_item_name: Optional[str]
    input_qty: Optional[Decimal]
    input_uom: Optional[str]
    input_supplier: Optional[str]
    output_lot_no: Optional[str]
    output_item_name: Optional[str]
    output_qty: Optional[Decimal]
    output_uom: Optional[str]
    link_type: str
    created_at: datetime


class GenealogyNode(BaseModel):
    order_id: str
    order_no: str
    batch_no: Optional[str]
    product_name: Optional[str]
    planned_qty: Decimal
    completed_qty: Decimal
    status: str
    is_rework: bool
    is_split: bool
    input_lots: List[BatchGenealogyOut]
    children: List["GenealogyNode"] = []

GenealogyNode.model_rebuild()


class GenealogyResult(BaseModel):
    root: GenealogyNode
    all_input_lots: List[str]
    all_output_lots: List[str]
    depth: int


# ── Production Order ───────────────────────────────────────────────────────────

class ProdExecOrderCreate(BaseModel):
    product_id: Optional[UUID] = None
    bom_id: Optional[UUID] = None
    routing_id: Optional[UUID] = None
    product_name: Optional[str] = None
    bom_code: Optional[str] = None
    planned_qty: Decimal
    uom: str = "KG"
    batch_no: Optional[str] = None
    source_type: str = "MANUAL"
    source_document_id: Optional[UUID] = None
    priority: int = 5
    planned_start: Optional[datetime] = None
    planned_end: Optional[datetime] = None
    plant_id: Optional[str] = None
    section_id: Optional[str] = None
    line_id: Optional[str] = None
    preferred_machine_id: Optional[UUID] = None
    remarks: Optional[str] = None


class ProdExecOrderSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    order_no: str
    product_id: Optional[UUID]
    product_name: Optional[str]
    bom_id: Optional[UUID]
    bom_code: Optional[str]
    planned_qty: Decimal
    completed_qty: Decimal
    scrap_qty: Decimal
    uom: str
    batch_no: Optional[str]
    source_type: str
    priority: int
    planned_start: Optional[datetime]
    planned_end: Optional[datetime]
    actual_start: Optional[datetime]
    actual_end: Optional[datetime]
    status: str
    is_rework: bool
    is_split: bool
    parent_order_id: Optional[UUID]
    total_actual_cost: Decimal
    cost_per_uom: Optional[Decimal]
    created_at: datetime
    updated_at: datetime


class ProdExecOrderOut(ProdExecOrderSummary):
    routing_id: Optional[UUID]
    plant_id: Optional[str]
    section_id: Optional[str]
    line_id: Optional[str]
    preferred_machine_id: Optional[UUID]
    planned_material_cost: Decimal
    actual_material_cost: Decimal
    planned_labor_cost: Decimal
    actual_labor_cost: Decimal
    overhead_cost: Decimal
    rework_qty: Decimal
    rework_reason: Optional[str]
    remarks: Optional[str]
    approved_by: Optional[UUID]
    work_orders: List[ExecWorkOrderOut] = []
    materials: List[ExecOrderMaterialOut] = []


# ── Action Requests ────────────────────────────────────────────────────────────

class CompleteOrderRequest(BaseModel):
    completed_qty: Decimal
    scrap_qty: Decimal = Decimal("0")
    remarks: Optional[str] = None


class SplitOrderRequest(BaseModel):
    split_qtys: List[Decimal]
    reason: Optional[str] = None


class MergeOrdersRequest(BaseModel):
    order_ids: List[UUID]
    reason: Optional[str] = None


class ReworkOrderRequest(BaseModel):
    qty: Decimal
    reason: str


class AddGenealogyRequest(BaseModel):
    input_lot_no: Optional[str] = None
    input_item_name: Optional[str] = None
    input_qty: Optional[Decimal] = None
    input_uom: Optional[str] = None
    input_supplier: Optional[str] = None
    output_lot_no: Optional[str] = None
    output_item_name: Optional[str] = None
    output_qty: Optional[Decimal] = None
    output_uom: Optional[str] = None
    link_type: str = "INPUT_OUTPUT"


# ── AI ─────────────────────────────────────────────────────────────────────────

class ExecAIRecOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    production_order_id: UUID
    agent_type: str
    status: str
    title: str
    explanation: Optional[str]
    impact_summary: Optional[str]
    confidence_score: Optional[Decimal]
    actioned_at: Optional[datetime]
    created_at: datetime


# ── Dashboard ──────────────────────────────────────────────────────────────────

class ExecDashboard(BaseModel):
    total_orders: int
    draft_orders: int
    released_orders: int
    in_progress_orders: int
    completed_today: int
    delayed_orders: int
    total_work_orders: int
    pending_work_orders: int
    active_work_orders: int
    total_scrap_qty: float
    pending_ai_recs: int
    by_status: dict
    recent_orders: List[ProdExecOrderSummary]


# ── Batch Record ───────────────────────────────────────────────────────────────

class BatchRecord(BaseModel):
    order: ProdExecOrderOut
    work_orders: List[ExecWorkOrderOut]
    materials: List[ExecOrderMaterialOut]
    genealogy: List[BatchGenealogyOut]
    generated_at: datetime

File: frontend/src/lib/productionExecution.ts
import { apiClient } from "./api";

// ── Enums ──────────────────────────────────────────────────────────────────────

export type ProdExecStatus =
  | "DRAFT" | "PLANNED" | "RELEASED" | "IN_PROGRESS"
  | "PAUSED" | "COMPLETED" | "CLOSED" | "CANCELLED";

export type WOExecStatus =
  | "PENDING" | "READY" | "IN_PROGRESS" | "PAUSED" | "COMPLETED" | "CANCELLED";

export type ExecSourceType =
  | "MRP" | "MPS" | "MANUAL" | "SALES" | "SUBCONTRACT_RETURN";

// ── Types ──────────────────────────────────────────────────────────────────────

export interface ExecWorkOrderOut {
  id: string;
  production_order_id: string;
  operation_code: string | null;
  step_sequence: number;
  step_name: string;
  work_center_id: string | null;
  section_id: string | null;
  line_id: string | null;
  machine_id: string | null;
  assigned_operator_id: string | null;
  planned_qty: number;
  completed_qty: number;
  scrap_qty: number;
  rework_qty: number;
  setup_time_planned: number;
  run_time_planned: number;
  cleanup_time_planned: number;
  setup_time_actual: number | null;
  run_time_actual: number | null;
  cleanup_time_actual: number | null;
  dependency_work_order_id: string | null;
  qc_required_flag: boolean;
  qc_passed: boolean | null;
  scrap_category: string | null;
  scrap_reason: string | null;
  status: WOExecStatus;
  started_at: string | null;
  completed_at: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface ExecOrderMaterialOut {
  id: string;
  production_order_id: string;
  work_order_id: string | null;
  item_type: string;
  item_id: string | null;
  item_name: string | null;
  item_code: string | null;
  required_qty: number;
  reserved_qty: number;
  issued_qty: number;
  consumed_qty: number;
  returned_qty: number;
  scrapped_qty: number;
  uom: string;
  status: string;
  lot_numbers: string[] | null;
  unit_cost: number | null;
  extended_cost: number | null;
}

export interface BatchGenealogyOut {
  id: string;
  production_order_id: string;
  input_lot_no: string | null;
  input_item_name: string | null;
  input_qty: number | null;
  input_uom: string | null;
  input_supplier: string | null;
  output_lot_no: string | null;
  output_item_name: string | null;
  output_qty: number | null;
  output_uom: string | null;
  link_type: string;
  created_at: string;
}

export interface ProdExecOrderSummary {
  id: string;
  order_no: string;
  product_id: string | null;
  product_name: string | null;
  bom_id: string | null;
  bom_code: string | null;
  planned_qty: number;
  completed_qty: number;
  scrap_qty: number;
  uom: string;
  batch_no: string | null;
  source_type: string;
  priority: number;
  planned_start: string | null;
  planned_end: string | null;
  actual_start: string | null;
  actual_end: string | null;
  status: ProdExecStatus;
  is_rework: boolean;
  is_split: boolean;
  parent_order_id: string | null;
  total_actual_cost: number;
  cost_per_uom: number | null;
  created_at: string;
  updated_at: string;
}

export interface ProdExecOrderOut extends ProdExecOrderSummary {
  routing_id: string | null;
  plant_id: string | null;
  section_id: string | null;
  line_id: string | null;
  preferred_machine_id: string | null;
  planned_material_cost: number;
  actual_material_cost: number;
  planned_labor_cost: number;
  actual_labor_cost: number;
  overhead_cost: number;
  rework_qty: number;
  rework_reason: string | null;
  remarks: string | null;
  approved_by: string | null;
  work_orders: ExecWorkOrderOut[];
  materials: ExecOrderMaterialOut[];
}

export interface ExecAIRecOut {
  id: string;
  production_order_id: string;
  agent_type: string;
  status: string;
  title: string;
  explanation: string | null;
  impact_summary: string | null;
  confidence_score: number | null;
  actioned_at: string | null;
  created_at: string;
}

export interface ExecDashboard {
  total_orders: number;
  draft_orders: number;
  released_orders: number;
  in_progress_orders: number;
  completed_today: number;
  delayed_orders: number;
  total_work_orders: number;
  pending_work_orders: number;
  active_work_orders: number;
  total_scrap_qty: number;
  pending_ai_recs: number;
  by_status: Record<string, number>;
  recent_orders: ProdExecOrderSummary[];
}

// ── Color maps ─────────────────────────────────────────────────────────────────

export const STATUS_COLOR: Record<ProdExecStatus, string> = {
  DRAFT:       "bg-gray-100 text-gray-600",
  PLANNED:     "bg-blue-100 text-blue-700",
  RELEASED:    "bg-indigo-100 text-indigo-700",
  IN_PROGRESS: "bg-yellow-100 text-yellow-700",
  PAUSED:      "bg-orange-100 text-orange-700",
  COMPLETED:   "bg-green-100 text-green-700",
  CLOSED:      "bg-teal-100 text-teal-700",
  CANCELLED:   "bg-red-100 text-red-700",
};

export const WO_STATUS_COLOR: Record<WOExecStatus, string> = {
  PENDING:     "bg-gray-100 text-gray-500",
  READY:       "bg-blue-100 text-blue-700",
  IN_PROGRESS: "bg-yellow-100 text-yellow-700",
  PAUSED:      "bg-orange-100 text-orange-700",
  COMPLETED:   "bg-green-100 text-green-700",
  CANCELLED:   "bg-red-100 text-red-700",
};

export const LIFECYCLE_NEXT: Partial<Record<ProdExecStatus, string>> = {
  DRAFT:       "Mark Planned",
  PLANNED:     "Release",
  RELEASED:    "Start",
  IN_PROGRESS: "Complete",
  COMPLETED:   "Close",
};

export function fmtKES(v: number): string {
  return `KES ${v.toLocaleString("en-KE", { minimumFractionDigits: 2, maximumFractionDigits: 2 })}`;
}

export function fmtMins(mins: number): string {
  const h = Math.floor(mins / 60);
  const m = Math.round(mins % 60);
  return h > 0 ? `${h}h ${m}m` : `${m}m`;
}

// ── API client ─────────────────────────────────────────────────────────────────

const BASE = "/production-execution";

export const execApi = {
  dashboard: () => apiClient.get<ExecDashboard>(`${BASE}/dashboard`),

  // Orders
  list:    (params?: { status?: string; product_id?: string }) => {
    const qs = new URLSearchParams(
      Object.entries(params || {}).filter(([, v]) => v) as [string, string][]
    ).toString();
    return apiClient.get<ProdExecOrderSummary[]>(`${BASE}${qs ? `?${qs}` : ""}`);
  },
  create:  (body: Partial<ProdExecOrderOut>) => apiClient.post<ProdExecOrderOut>(BASE, body),
  get:     (id: string) => apiClient.get<ProdExecOrderOut>(`${BASE}/${id}`),
  update:  (id: string, body: Partial<ProdExecOrderOut>) => apiClient.patch<ProdExecOrderOut>(`${BASE}/${id}`, body),

  // Lifecycle
  advance:  (id: string) => apiClient.post<ProdExecOrderOut>(`${BASE}/${id}/advance`, {}),
  release:  (id: string) => apiClient.post<ProdExecOrderOut>(`${BASE}/${id}/release`, {}),
  start:    (id: string) => apiClient.post<ProdExecOrderOut>(`${BASE}/${id}/start`, {}),
  pause:    (id: string) => apiClient.post<ProdExecOrderOut>(`${BASE}/${id}/pause`, {}),
  complete: (id: string, body: { completed_qty: number; scrap_qty?: number; remarks?: string }) =>
    apiClient.post<ProdExecOrderOut>(`${BASE}/${id}/complete`, body),
  close:    (id: string) => apiClient.post<ProdExecOrderOut>(`${BASE}/${id}/close`, {}),
  cancel:   (id: string) => apiClient.post<ProdExecOrderOut>(`${BASE}/${id}/cancel`, {}),

  // Split / Merge / Rework
  split:  (id: string, split_qtys: number[], reason?: string) =>
    apiClient.post<ProdExecOrderOut[]>(`${BASE}/${id}/split`, { split_qtys, reason }),
  merge:  (order_ids: string[], reason?: string) =>
    apiClient.post<ProdExecOrderOut>(`${BASE}/merge`, { order_ids, reason }),
  rework: (id: string, qty: number, reason: string) =>
    apiClient.post<ProdExecOrderOut>(`${BASE}/${id}/rework`, { qty, reason }),

  // Genealogy
  genealogy: (id: string) => apiClient.get<any>(`${BASE}/${id}/genealogy`),
  addGenealogyLink: (id: string, body: Partial<BatchGenealogyOut>) =>
    apiClient.post<BatchGenealogyOut>(`${BASE}/${id}/genealogy`, body),

  // Batch record
  batchRecord: (id: string) => apiClient.get<any>(`${BASE}/${id}/batch-record`),

  // Work Orders
  listWorkOrders:    (params?: { status?: string }) => {
    const qs = params?.status ? `?status=${params.status}` : "";
    return apiClient.get<ExecWorkOrderOut[]>(`${BASE}/work-orders${qs}`);
  },
  addWorkOrder:     (orderId: string, body: Partial<ExecWorkOrderOut>) =>
    apiClient.post<ExecWorkOrderOut>(`${BASE}/${orderId}/work-orders`, body),
  startWorkOrder:   (woId: string) =>
    apiClient.post<ExecWorkOrderOut>(`${BASE}/work-orders/${woId}/start`, {}),
  completeWorkOrder:(woId: string, body: Record<string, unknown>) =>
    apiClient.post<ExecWorkOrderOut>(`${BASE}/work-orders/${woId}/complete`, body),
  updateWorkOrder:  (woId: string, body: Partial<ExecWorkOrderOut>) =>
    apiClient.patch<ExecWorkOrderOut>(`${BASE}/work-orders/${woId}`, body),

  // Materials
  addMaterial:    (orderId: string, body: Partial<ExecOrderMaterialOut>) =>
    apiClient.post<ExecOrderMaterialOut>(`${BASE}/${orderId}/materials`, body),
  issueMaterial:  (orderId: string, body: { material_id: string; qty: number; lot_no?: string; action: string }) =>
    apiClient.post<ExecOrderMaterialOut>(`${BASE}/${orderId}/materials/issue`, body),
  returnMaterial: (orderId: string, body: { material_id: string; qty: number; action: string }) =>
    apiClient.post<ExecOrderMaterialOut>(`${BASE}/${orderId}/materials/return`, body),

  // AI
  runAI:    (id: string) => apiClient.post<{ total: number }>(`${BASE}/${id}/run-ai`, {}),
  listAI:   (id: string) => apiClient.get<ExecAIRecOut[]>(`${BASE}/${id}/ai-recs`),
  actionAI: (recId: string, status: string) =>
    apiClient.post<ExecAIRecOut>(`${BASE}/ai-recs/${recId}/action`, { status }),
};
